x-pack/plugins/enterprise_search/server/lib/enterprise_search_config.ts

```diff
@@ -0,0 +1,134 @@
+
+export const config: PluginConfigDescriptor<EnterpriseSearchConfigType> = {
+  schema: configSchema,
+  exposeToBrowser: {
+    host: true,
+  },
+};
+
+// -----------------------------------------------------------------------------
+/**
+ * Wrapper of config schema.
+ * @public
+ */
+export class EnterpriseSearchConfig {
+  /**
+   * Base URL (schema://host:port) of the Enterprise Search deployment
+   */
+  public host: string;
+
+  /**
+   * Specifies if Enterprise Search plugin should be enabled
+   */
+  public enabled: boolean;
+
+  /**
+   * Specifies the timeout for all API calls to Enterprise Search
+   */
+  public accessCheckTimeout: number;
+
+  /**
+   * Specifies a warning threshold for all API calls to Enterprise Search.
+   * If a request takes longer than this value, a user will be presented
+   * with a warning message asking them to make sure everything is OK.
+   */
+  public accessCheckTimeoutWarning: number;
+
+  /**
+   * A list of custom certificate authorities to be used for connecting to
+   * Enterprise Search APIs. If empty, default list of root CAs is used.
+   */
+  public certificateAuthorities: string[];
+
+  /**
+   * If set to true, SSL errors when connecting to Enterprise Search will be
+   * ignored as long as a successful SSL handshake could be performed.
+   */
+  public rejectUnauthorized: boolean;
+
+  /**
+   * HTTP agent object to be used for all Enterprise Search API calls
+   */
+  public httpAgent: http.Agent | https.Agent;
+
+  /**
+   * @internal
+   */
+  constructor(rawConfig: EnterpriseSearchConfigType) {
+    this.host = rawConfig.host;
+    this.enabled = rawConfig.enabled;
+    this.accessCheckTimeout = rawConfig.accessCheckTimeout;
+    this.accessCheckTimeoutWarning = rawConfig.accessCheckTimeoutWarning;
+    this.rejectUnauthorized = rawConfig.ssl.rejectUnauthorized;
+    this.certificateAuthorities = this.loadCertificateAuthorities(
+      rawConfig.ssl.certificateAuthorities
+    );
+    this.httpAgent = this.httpAgentFor(this.host);
+  }
+
+  /**
+   * Loads custom CA certificate files and returns all certificates as an array
+   **/
+  private loadCertificateAuthorities(ca) {
+    const parsedCerts = [];
+    if (ca) {
+      const paths = Array.isArray(ca) ? ca : [ca];
+      if (paths.length > 0) {
+        for (const path of paths) {
+          const parsedCert = readFileSync(path, 'utf8');
+          parsedCerts.push(parsedCert);
+        }
+      }
+    }
+    return parsedCerts;
+  }
+
+  /**
+   * Returns an HTTP agent object to be used for connecting to a given host
+   */
+  private httpAgentFor(host) {
+    try {
+      const parsedHost = new URL(host);
+      if (parsedHost.protocol === 'https:') {
+        return new https.Agent({
+          ca: this.certificateAuthorities,
+          rejectUnauthorized: this.rejectUnauthorized,
+        });
+      }
+    } catch (_error) {
+      // ignore the error
+    }
+
+    // Use HTTP agent by default
+    return new http.Agent();
```

Comment:
[Noob question] Does it matter if a server is using `https` but doesn't have the `ssl` or `ssl.certificateAuthorities` config setting set?

Sorry, trying to rephrase this in a better way... essentially, do we need to check for `https://` vs `http://` protocal on **all** `config.host` URLs (regardless of whether the new `ssl` config is set) and return a default `https.Agent()` or `http.Agent()` depending on the protocol?

Or will `http.Agent()` work for both http and https hosts?

Reply:
How would we end up in a place where the schema used for `enterpriseSearch.host` would differ from the schema of URLs used to call Enterprise Search?

Reply:
ha, good question. No idea, just wasn't sure if `http.Agent` would work as-is if both Enterprise Search and Kibana were on `https`.

Reply:
New unit tests gave me a great way of re-asking my question in a more clear manner 😄 https://github.com/elastic/kibana/pull/100946#discussion_r643397211 - I'll resolve this and continue the conversation there